subprojects/core-api/src/main/java/org/gradle/api/model/ObjectFactory.java

```diff
@@ -104,6 +105,25 @@
      */
     <T> Property<T> property(Class<T> valueType);
 
+    /**
+     * Creates a {@link Property} implementation to hold values of the given type. If nothing else
+     * sets this property value, it will be computed from the supplied callable.
+     *
+     * <p>For certain types, there are more specialized property factory methods available:</p>
+     * <ul>
+     * <li>For {@link List} properties, you should use {@link #listProperty(Class)}.</li>
+     * <li>For {@link Set} properties, you should use {@link #setProperty(Class)}.</li>
+     * <li>For {@link org.gradle.api.file.Directory} properties, you should use {@link #directoryProperty()}.</li>
+     * <li>For {@link org.gradle.api.file.RegularFile} properties, you should use {@link #fileProperty()}.</li>
+     * </ul>
+     *
+     * @param valueType The type of the property.
+     * @param defaultValue the default value generator
+     * @return The property. Never returns null.
+     * @since 5.1
+     */
+    <T> Property<T> property(Class<T> valueType, Callable<? extends T> defaultValue);
```

Comment:
There's already a way to give an initial value for a `Property<T>` given an instance of `T` by chaining a call to `value()`, e.g. `my prop = property(type).value(x)`. We should try to use the same pattern to define an initial value given something that can calculate a `T`. That would mean either adding a `value(Provider<T>)` method to `Property` instead of this factory method, or it would mean adding more overloads to this factory interface and also removing `value(T)` from `Property`. Personally, I'd rather we add more method to `Property` rather than a bunch of overloaded factory methods.

I think we should also strongly encourage people to use `Provider<T>` instead of `Callable<T>` when they need to provide a calculated value, because `Callable` is completely opaque to us. For example, it doesn't carry any information about the tasks the produce the value, or even what type of value it produces.

Reply:
Right. I'm not super familiar with the `Provider<T>` API but my issue is two fold:

1. as you noticed, `value(T)` is eager, so it sets an initial value, but it has to be calculated eagerly, so I cannot derive it from other configuration
2. I'm fine with `value(Provider<T>)`, but then it just defers the problem a little more: how am I supposed to create a `Provider<T>` from something that can calculate a `T`? The only API I see is `providerFactory.provider(Callable<T>)`. Is this fine? If so, would adding `property(type).provider(Provider<T>))` be acceptable? It would still be a bit verbose, but it would work:

```
final Property<String> myProp = project.getObjects().property(String.class).provider(project.getProviders().provider( () -> ...)
```

This is still far more verbose and less discoverable than I would like, but I see your point.

Reply:
If we do add such overload, we need to be careful as to how users are using it and how the implicit task dependencies are transferred between properties especially the one inside the `Callable`. We want to make sure users aren't unpacking provider manually which result in losing the task dependencies.